banzai/images.py

```diff
@@ -1,272 +1,766 @@
-import os
 import logging
-import tempfile
-import shutil
-import datetime
-
+from banzai import dbs
+from banzai.utils import fits_utils, stats, date_utils
 import numpy as np
 from astropy.io import fits
-from astropy.table import Table, Column
-
-import banzai
-from banzai import dbs, settings, exceptions
-from banzai.utils import date_utils, file_utils, fits_utils
-from banzai import logs
+from astropy.table import Table
+from astropy.time import Time
+from typing import Union, Type
+import tempfile
+import abc
+import os
 
 logger = logging.getLogger('banzai')
 
 
-class DataTable(object):
-    """
-    Object for storing astropy (or another table type) tables with an additional .name attribute which
-    sets the table's extension name when it is saved as a fits file.
-    """
-    def __init__(self, data_table, name):
+class Data(metaclass=abc.ABCMeta):
+    _file_handles = []
+
+    def __init__(self, data: Union[np.array, Table], meta: Union[dict, fits.Header],
+                 mask: np.array = None, name: str = ''):
+        self.data = self._init_array(data)
+        self.meta = meta
+        self._validate_mask(mask)
+        self.mask = self._init_array(mask, dtype=np.uint8)
         self.name = name
-        self._data_table = data_table
 
-    def __getitem__(self, item):
-        return self._data_table[item]
+    def _validate_mask(self, mask):
+        if mask is not None:
+            if mask.shape != self.data.shape:
+                raise ValueError('Mask must have the same dimensions as the data')
+
+    def _init_array(self, array: np.array = None, dtype: Type = None):
+        file_handle = tempfile.NamedTemporaryFile('w+b')
+        if array is None:
+            shape = self.data.shape
+            if dtype is None:
+                dtype = self.data.dtype
+            array = np.zeros(shape, dtype=dtype)
+        if array.size > 0:
+            memory_mapped_array = np.memmap(file_handle, shape=array.shape, dtype=array.dtype, mode='readwrite')
+            memory_mapped_array.ravel()[:] = array.ravel()[:]
+            self._file_handles.append(file_handle)
+        else:
+            memory_mapped_array = array
+        return memory_mapped_array
 
-    def __setitem__(self, key, value):
-        self._data_table[key] = value
+    def add_mask(self, mask: np.array):
+        self._validate_mask(mask)
+        self.mask = self._init_array(mask)
 
-    def add_column(self, arr, name, index=None):
-        self._data_table.add_column(Column(arr), name=name, index=index)
+    def __del__(self):
+        for handle in self._file_handles:
+            handle.close()
+        del self.data
+        del self.mask
 
```

Comment:
Do we need to override the standard delete method? When we do del CCDdata_instance, does it not delete every attribute it has from memory?